components/brave_wallet/browser/internal/orchard_block_scanner.cc

```diff
@@ -29,50 +33,57 @@ OrchardBlockScanner::~OrchardBlockScanner() = default;
 
 base::expected<OrchardBlockScanner::Result, OrchardBlockScanner::ErrorCode>
 OrchardBlockScanner::ScanBlocks(
-    std::vector<OrchardNote> known_notes,
+    OrchardTreeState tree_state,
     std::vector<zcash::mojom::CompactBlockPtr> blocks) {
+  std::unique_ptr<orchard::OrchardDecodedBlocksBundle> result =
+      decoder_->ScanBlocks(tree_state, blocks);
+  if (!result) {
+    DVLOG(1) << "Failed to parse block range.";
+    return base::unexpected(ErrorCode::kInputError);
+  }
+
+  if (!result->GetDiscoveredNotes()) {
+    DVLOG(1) << "Failed to resolve discovered notes.";
+    return base::unexpected(ErrorCode::kInputError);
+  }
+
   std::vector<OrchardNoteSpend> found_spends;
-  std::vector<OrchardNote> found_notes;
+  std::vector<OrchardNote> found_notes = result->GetDiscoveredNotes().value();
 
   for (const auto& block : blocks) {
-    // Scan block using the decoder initialized with the provided fvk
-    // to find new spendable notes.
-    auto scan_result = decoder_->ScanBlock(block);
-    if (!scan_result) {
-      return base::unexpected(ErrorCode::kDecoderError);
-    }
-    found_notes.insert(found_notes.end(), scan_result->begin(),
-                       scan_result->end());
-    // Place found notes to the known notes list so we can also check for
-    // nullifiers
-    known_notes.insert(known_notes.end(), scan_result->begin(),
-                       scan_result->end());
     for (const auto& tx : block->vtx) {
       // We only scan orchard actions here
       for (const auto& orchard_action : tx->orchard_actions) {
         if (orchard_action->nullifier.size() != kOrchardNullifierSize) {
           return base::unexpected(ErrorCode::kInputError);
         }
 
-        std::array<uint8_t, kOrchardNullifierSize> action_nullifier;
-        base::ranges::copy(orchard_action->nullifier, action_nullifier.begin());
-
+        OrchardNoteSpend spend;
         // Nullifier is a public information about some note being spent.
         // Here we are trying to find a known spendable notes which nullifier
         // matches nullifier from the processed transaction.
-        if (std::find_if(known_notes.begin(), known_notes.end(),
-                         [&action_nullifier](const auto& v) {
-                           return v.nullifier == action_nullifier;
-                         }) != known_notes.end()) {
-          OrchardNoteSpend spend;
-          spend.block_id = block->height;
-          spend.nullifier = action_nullifier;
-          found_spends.push_back(std::move(spend));
-        }
+
+        base::ranges::copy(orchard_action->nullifier, spend.nullifier.begin());
+        spend.block_id = block->height;
+        found_spends.push_back(std::move(spend));
       }
     }
   }
-  return Result({std::move(found_notes), std::move(found_spends)});
+
+  return Result(
+      {std::move(found_notes), std::move(found_spends), std::move(result)});
```

Comment:
`std::move(found_notes).value()` if you change `found_notes` to an optional vector.